src/main/java/net/snowflake/ingest/connection/JWTManager.java

```diff
@@ -0,0 +1,181 @@
+    // if any of our arguments are null, throw an exception
+    if (keyPair == null) {
+      throw new IllegalArgumentException();
+    }
+    token = new AtomicReference<>();
+
+    // we have to keep around the keys
+    this.keyPair = keyPair;
+
+    // generate our first token
+    regenerateToken();
+
+    // schedule all future renewals
+    tokenRefresher.scheduleAtFixedRate(
+        this::regenerateToken, timeTillRenewal, timeTillRenewal, unit);
+  }
+
+  /**
+   * Creates a JWTManager entity for a given account, user and KeyPair with the default time to
+   * renew (RENEWAL_INTERVAL_IN_MINUTES minutes)
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param keyPair - the public/private key pair we're using to connect
+   * @param telemetryService reference to the telemetry service
+   */
+  JWTManager(
+      String accountName, String username, KeyPair keyPair, TelemetryService telemetryService) {
+    this(
+        accountName,
+        username,
+        keyPair,
+        RENEWAL_INTERVAL_IN_MINUTES,
+        TimeUnit.MINUTES,
+        telemetryService);
+  }
+
+  @Override
+  String getToken() {
+    // if we failed to regenerate the token at some point, throw
+    if (refreshFailed.get()) {
+      LOGGER.error("getToken request failed due to token regeneration failure");
+      throw new SecurityException();
+    }
+
+    return token.get();
+  }
+
+  @Override
+  String getTokenType() {
+    return TOKEN_TYPE;
+  }
+
+  /** regenerateToken - Regenerates our Token given our current user, account and keypair */
+  private void regenerateToken() {
+    // create our JWT claim builder object
+    JWTClaimsSet.Builder builder = new JWTClaimsSet.Builder();
+
+    // get the subject to the fully qualified username
+    String subject = String.format("%s.%s", account, user);
+
+    // get the issuer
+    String publicKeyFPInJwt = calculatePublicKeyFp(keyPair);
+    String issuer = String.format("%s.%s.%s", account, user, publicKeyFPInJwt);
+
+    // iat set to now
+    Date iat = new Date(System.currentTimeMillis());
+
+    // expiration in 59 minutes
+    Date exp = new Date(iat.getTime() + 59 * 60 * 1000);
+
+    // build claim set
+    JWTClaimsSet claimsSet =
+        builder.issuer(issuer).subject(subject).issueTime(iat).expirationTime(exp).build();
+    LOGGER.debug("Creating new JWT with subject {} and issuer {}...", subject, issuer);
+
+    SignedJWT signedJWT = new SignedJWT(new JWSHeader(JWSAlgorithm.RS256), claimsSet);
+
+    JWSSigner signer = new RSASSASigner(this.keyPair.getPrivate());
+
+    String newToken;
+    try {
+      signedJWT.sign(signer);
+      newToken = signedJWT.serialize();
+    } catch (JOSEException e) {
+      refreshFailed.set(true);
+      LOGGER.error("Failed to regenerate token! Exception is as follows : {}", e.getMessage());
+      throw new SecurityException();
```

Comment:
Codecov is complaining here.

Reply:
This part of code existed in older version and could not be covered unless adding injection function.